buildSrc/src/main/kotlin/VersionFile.kt

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2016-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+import org.gradle.api.*
+import org.gradle.api.tasks.*
+
+/**
+ * Adds 'module_name.version' file to the project's JAR META-INF
+ * for the better toolability. See #2941
+ */
+object VersionFile {
+    fun registerVersionFileTask(project: Project): TaskProvider<Task> = with(project) {
+        tasks.register("versionFileTask") {
+            val name = project.name.replace('-', '_')
+            val versionFile = project.layout.buildDirectory.file("$name.version")
+            outputs.file(versionFile)
+            doLast {
+                versionFile.get().asFile.writeText(project.version.toString())
```

Comment:
Use of `project` in task execution phase breaks configuration cache compatibility.
Assign the `version` to a variable outside of the `doLast` block to avoid that, just like you did with `name`.

Probably, `with(project)` will also break it (unsure), it's redundant here your could to `project.tasks.register`.